### PR TITLE
treewide/nixos: remove `with lib;` part 3

### PR DESCRIPTION
**Pull request by Stunkymonkey**
## Description of changes

part of #208242

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Stunkymonkey** [comment]
@philiptaron how should we continue here? do we consider my script as too bad?

**Comment by philiptaron** [comment]
> @philiptaron how should we continue here? do we consider my script as too bad?

The fallout has made me pretty reticent to continue. It takes me about an hour to review these, and I'm provably fallible with the defects that have slipped through. I also have to switch over to $work$ for a bit, so I'm a bit slower than I was before.

**Review by Ma27 — DISMISSED** [review]
AFAIK this is done by some automation: https://github.com/NixOS/nixpkgs/pull/335603#issuecomment-2326936584

Given that this has always caused a notable fallout that maintainers had to fix afterwards, did you fix the script or are we expecting a different result this time?

**Comment by Stunkymonkey** [comment]
> AFAIK this is done by some automation: [#335603 (comment)](https://github.com/NixOS/nixpkgs/pull/335603#issuecomment-2326936584)
> 
> Given that this has always caused a notable fallout that maintainers had to fix afterwards, did you fix the script or are we expecting a different result this time?

yes exactly the same script. Automation is not changed.

I was expecting 2 reviewers is enough to detect the problems. Definitely not a fault of the reviewers!

**Comment by Stunkymonkey** [comment]
@Ma27 I have implemented a new approach:

using `nix-build nixos/release.nix -A manual.x86_64-linux` to build and parse the output of it with a custom python script. So i am only correcting the reported errors. (based on line-, character-index reports.)

I am doing this until the build succeeds. Yes it is slow, but it surprisingly works:

<details><summary>python-script</summary>

```python
#!/usr/bin/env python3
import subprocess
import re
import sys


def get_option_string():
    result = subprocess.run(
        [
            "nix",
            "eval",
            "--impure",
            "--expr",
            'with import <nixpkgs> {}; builtins.attrNames (builtins.removeAttrs lib ["__unfix__" "removeAttrs" "map" "licenses" "fix" "and" "or" "kernel" "options" "meta" "path"])',
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    return result.stdout.replace('" "', " ").replace('" ]', "").replace('[ "', "")


def main():
    print("start post-patch")
    optionstring = get_option_string()
    available_lib_options = set(optionstring.split())
    NIX_BUILD_CMD = "nix-build nixos/release.nix -A manual.x86_64-linux"

    for file_path in sys.argv[1:]:
        while True:
            result = subprocess.run(
                NIX_BUILD_CMD.split(),
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
            error_output = result.stderr
            reversed_error = error_output.splitlines()[::-1]
            error_variable_match = re.search(
                r"undefined variable '(\w+)'", error_output
            )
            # print(error_output, error_variable_match)
            if error_variable_match:
                error_variable = error_variable_match.group(1)
                line_info = next(
                    (line for line in reversed_error if "  at " in line), None
                )
                if line_info and error_variable in available_lib_options:
                    print(f"loc: {line_info.strip()}")
                    line_number = (
                        int(re.search(r":([0-9]+):[0-9]+:", line_info).group(1)) - 1
                    )
                    char_number = (
                        int(re.search(r":[0-9]+:([0-9]+):", line_info).group(1)) - 1
                    )
                    with open(file_path, "r") as file:
                        lines = file.readlines()
                    error_line = lines[line_number]
                    print(f"old: {error_line.strip()}")
                    modified_line = (
                        error_line[:char_number] + "lib." + error_line[char_number:]
                    )
                    print(f"new: {modified_line.strip()}")
                    lines[line_number] = modified_line
                    with open(file_path, "w") as file:
                        file.writelines(lines)
            else:
                break


if __name__ == "__main__":
    main()
```

</details>

is this a better approach? are you fine with it?

**Comment by Ma27** [comment]
> is this a better approach? are you fine with it?

I think it's a better approach than what we had before.
Another idea: evaluate relevant VM tests (if any): this approach covers option declarations and stuff around those, but I think e.g. code that puts systemd untis together isn't affected?
It may also help to do `<module>.enable = true;`, evaluate and check if something happens that isn't a ("X is used but not defined")?

That being said, my main concern was that seemingly there didn't happen much QA during the past iterations and we somehow expected different results on followups, hence my first comment.

But, after going through the options, I don't really have a good answer on how to test these kinds of changes.
@Infinisil or @roberth you perhaps?

**Comment by Stunkymonkey** [comment]
@Ma27 should we split each commit into a separate MR? this way it would take longer, but the separation is better.

**Comment by philiptaron** [comment]
I'm going to give it a human review and see the error rate that I can find.

**Review by philiptaron — APPROVED** [review]
I can't find a single error of commission in this PR, meaning that (a) the error rate for this new script is much lower and (b) whatever errors this script produces are going to be errors of omission rather than commission.

I'm willing to merge it given the review.

In particular, given this line in the script:

```
with import <nixpkgs> {};
builtins.attrNames (builtins.removeAttrs lib ["__unfix__" "removeAttrs" "map" "licenses" "fix" "and" "or" "kernel" "options" "meta" "path"])
```

The likely errors are places where `licenses`, `fix`, `and`, `or`, `kernel`, `options`, `meta`, or `path` are actually used in a NixOS module, as the script will skip them. `lib.or` is such a strange beast that I am going to discount that one.

Any file which does not contain these tokens will not have an error, of that I am fairly certain. If a file _does_ contain these tokens, the context needs to be taken into account -- most of the uses are unlikely to be errors.

File: nixos/modules/services/amqp/activemq/default.nix
[the remark this answers is not in the document]
Comment (Stunkymonkey) [reply nixos/modules/services/amqp/activemq/default.nix:1]: the best part about it: even less work for me...

will rebase and start working on the next part

File: nixos/modules/services/misc/paperless.nix [hunk 1]
```diff
@@ -229,29 +227,11 @@ in
 
       This sets `OMP_NUM_THREADS` to `1` in order to mitigate the issue. See
       https://github.com/NixOS/nixpkgs/issues/240591 for more information
-    '' // mkOption { default = true; };
```
Comment (philiptaron) [inline nixos/modules/services/misc/paperless.nix:232]: this isn't expected, right, @Stunkymonkey?

Comment (Stunkymonkey) [reply nixos/modules/services/misc/paperless.nix:232]: you are totally right, might be i did something wrong while rebasing.

I reverted all my changes and rerun the script.

Comment (philiptaron) [reply nixos/modules/services/misc/paperless.nix:232]: On this file only, I hope!

File: nixos/modules/services/misc/redlib.nix [hunk 2]
```diff
@@ -88,8 +87,7 @@ in
     };
   };
 
-  config = mkIf cfg.enable {
-    systemd.packages = [ cfg.package ];
```
Comment (philiptaron) [inline nixos/modules/services/misc/redlib.nix:92]: Here's another rebase error, I'm assuming.

Comment (Stunkymonkey) [reply nixos/modules/services/misc/redlib.nix]: removed the commit touching that file. already resolved via e286b91ebca6f7de81cefb87492440582a565990

**Review by philiptaron — APPROVED** [review]

**Comment by philiptaron** [comment]
Let's ship it and see what shakes out.

**Comment by philiptaron** [comment]
2 days later and no error reports, just a rebase ping. Way to go Felix!

**Pull request merged**